go/vt/vtgate/tx_conn.go

```diff
@@ -78,30 +80,47 @@ func (txc *TxConn) Commit(ctx context.Context, session *SafeSession) error {
 	return txc.commitNormal(ctx, session)
 }
 
+func (txc *TxConn) commitShard(ctx context.Context, s *vtgatepb.Session_ShardSession) error {
+	var qs queryservice.QueryService
+	var err error
+	if UsingLegacyGateway() {
+		qs = txc.gateway
+	} else {
+		qs, err = txc.gateway.QueryServiceByAlias(s.TabletAlias)
+		if err != nil {
+			return err
+		}
+	}
+	// TODO(deepthi): TransactionId should be set to 0 only if the commit succeeded
```

Comment:
I don't feel this is correct. If we tried and failed to commit, we should not trust that we still have an open transaction on this connection. I think we should set the tx_id to 0 and close the connection, since we can't trust it

Reply:
If there's an error, we rollback and that resets the session (which I assume is effectively "closing" the connection). If the commit is successful also should set the TransactionID to 0 to denote that the transaction has ended. It's the defer call that is problematic IMO. The way I think this should work is:
1. Commit is successful. All shardSession transactions have ended, so set all TransactionID = 0
2. There was an error from one or more shards. Rollback the _other_ shards. Rollback should only operate on the shards that did not fail. So set TransactionID=0 on failed shards so that we attempt rollback only on successful shards. There is no need to explicitly set TransactionID = 0 on the successful shards at the end of rollback. the defer call to session.Reset takes care of that.

Reply:
SGTM